### ISSUE TITLE
Set esModuleInterop to "true" by default

### ISSUE THREAD

**Issue by voxpelli** [microsoft/TypeScript#41139]
## Search Terms

* esModuleInterop

## Suggestion

Change [`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) from something being an opt-in to instead be an opt-out, taking one step further towards deprecating the non-`esModuleInterop` mode and make `esModuleInterop` the only supported mode going forward (which has to be the long term goal).

## Use Cases

When [Typescript 2.7 introduced](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#support-for-import-d-from-cjs-from-commonjs-modules-with---esmoduleinterop) `esModuleInterop` almost 3 years ago it emphasized clearly that:

> We highly recommend applying it both to new and existing projects.

[`esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) is also a _recommended_ option to set in `tsconfig.json` and when one runs `tsc --init` it gets set to `true` automatically.

However, the fact that it still is opt-in makes that non-`esModuleInterop` mode is still very prevalent, which kind of works against the very problems it was meant to fix – as rather than having just the pre-`esModuleInterop` functionality, modules now have to try to make things work across both options – and on top of that increasingly have to work with TS-validated JS.

All in all, that causes problems: https://github.com/fastify/env-schema/pull/17

And is hard to wrap ones head around: https://github.com/fox1t/modules-playground

Even if one really tries: https://github.com/fox1t/modules-playground/pull/3

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
  * It would for all code which haven't yet adopted `esModuleInterop`, but it would make it easier to get TypeScript and JavaScript code to stay compatible with one another
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
  * It would emit different JS for all code which haven't yet adopted `esModuleInterop`, but it would make it easier to get TypeScript and JavaScript code to stay compatible with one another
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).
  * Apart from `Do not cause substantial breaking changes from TypeScript 1.0.`

**Comment by RyanCavanaugh** [comment]
I'll take this as a possibility but we have historically not changed commandline defaults unless the case for it was overwhelmingly strong. This might have the possibility to break people's programs and we would need some very good evidence that it was worth those people encountering that problem.

**Comment by voxpelli** [comment]
@RyanCavanaugh Yeah, it's tricky, right now it's causing resentment among non-TS developers who are trying to cater TS users, so current situation is undesirable as well. (See eg also this Twitter thread with me and @mcollina on top of the above GitHub issues: https://twitter.com/matteocollina/status/1305953416217268226)

**Comment by voxpelli** [comment]
An example where the current default causes confusion and makes people stick with the older non-recommended path: https://github.com/plantain-00/type-coverage/issues/78

Also had this Twitter conversation with @rich-harris: https://twitter.com/rich_harris/status/1361839944453541892?s=21

**Comment by bmeck** [comment]
I'm a bit wary of using `esModuleInterop` since how it interacts with node's ESM implementation of `default` isn't quite compatible. This leads to various conversations about what to do usually ending up with "don't use default exports" as my go to response. Enabling it by default would remove some issues but it seems like a second migration would have to happen afterward if node compatibility was another configuration option.

**Comment by voxpelli** [comment]
> I'm a bit wary of using `esModuleInterop` since how it interacts with node's ESM implementation of `default` isn't quite compatible.

@bmeck Can you elaborate a bit on this or provide a link which describes it?

My impression was that `esModuleInterop` was meant to increase compatibility?

**Comment by bmeck** [comment]
@voxpelli

> My impression was that esModuleInterop was meant to increase compatibility?

It kind of does, but not enough to be completely compatible. Particularly how `"default"` works still tries to do some magic and things start to get incompatible because of it.

Given:

```ts
// ./entry.ts
import * as asNamespace from './dep.js';
import asDefault from './dep.js';

console.log({
  asDefault,
  asNamespace
});
```

```cjs
// ./dep.js
module.exports = {
  __esModule: true,
  default: 'prop named default'
};
```

Compiling to CommonJS will result in the following outputs:

Node (after renaming `.ts` to `.mjs`):

```console
$ node ./entry.mjs
{
  asDefault: { __esModule: true, default: 'prop named default' },
  asNamespace: [Module: null prototype] {
    __esModule: true,
    default: { __esModule: true, default: 'prop named default' }
  }
}
```

After TS:

```console
$ node ./out/entry.js
{ asDefault: 'prop named default',
  asNamespace: { __esModule: true, default: 'prop named default' } }
```

Namely, even with `esModuleInterop` it tries to do some hoisting of the `default` property if `__esModule:true` exists that Node won't ever do.

**Comment by voxpelli** [comment]
The #44501 seems to partially implement this as both its module types imply `esModuleInterop`

So anyone setting the [`'module'`](https://www.typescriptlang.org/tsconfig#module) in their `tsconfig.json` to either `node12` and `nodenext` will get `esModuleInterop` by default.

A good solution for changing the default without breaking backwards compatibility, but may complicate the uptake of those new types, see eg: https://github.com/sindresorhus/tsconfig/issues/8

**Comment by voxpelli** [comment]
> The #44501 seems to partially implement this as both its module types imply `esModuleInterop`
> 
> So anyone setting the [`'module'`](https://www.typescriptlang.org/tsconfig#module) in their `tsconfig.json` to either `node12` and `nodenext` will get `esModuleInterop` by default.

Double checked whether this actually shipped, and it did: https://github.com/microsoft/TypeScript/blob/2bcfed01f3458996e71ce37af43e3495cb7e4950/src/compiler/utilities.ts#L6428-L6438

So anyone using `"module": "node16"` or `"module": "nodenext"` will have their `esModuleInterop` default to `true` (but can override it by actively setting it to `false`).

Though: This isn't at all mentioned in [the docs for `esModuleInterop`](https://www.typescriptlang.org/tsconfig#esModuleInterop) 🤔
